### PR TITLE
The great toxification (part 1)

### PR DESCRIPTION
**Pull request by stephenfin**
Subject: Make tox a little more friendly

Make `tox` a little more friendly by enabling extras like coverage, test profiling etc. by default, and closing gaps with the current Makefile targets such as PyLint and building docs in particular formats.

### Feature or Bugfix
- Feature

### Purpose
To paraphrase the Zen of Python (import this):

> tox is one honking great idea -- let's use more of it!

Start modernizing the test infrastructure that's using make and other hand-crafted tools by replacing them with tox-based derivatives.

**This is part one.**

### Detail
This is a multi-step process, with the end goal of completely removing the Makefile and running everything through `tox`. The first fours steps of these can be done in parallel, but combined they block the final two.

1. Make `tox` a little more friendly by enabling extras like coverage, test profiling etc. by default, and closing gaps with the current Makefile targets such as PyLint and building docs in particular formats (**this change**)
2. Replace/remove custom tooling found in `utils` in favor of modern alternatives like `flake8`
3. Centralize as much configuration as possible inside `setup.cfg` and `tox.ini`, to avoid polluting the top level directory.
4. Clean up requirements (`test-reqs.txt`, `setup.py`) to ensure virtualenv and non-virtualenv builds are identical

The remaining two steps are blocked by the above:

5. Switch CIs to use `tox` instead of the Makefile, and update `CONTRIBUTING.rst` to refer to tox. Deprecate the Makefile
6. (Bonus) Integrate [codecov.io](https://codecov.io) to make use of the coverage stats we now have

### Relates
- N/A

### CODE REVIEW

**Review by tk0miya — APPROVED** [review]
+1!

File: tox.ini [hunk 1]
```diff
@@ -6,15 +6,16 @@ passenv = https_proxy http_proxy no_proxy
...
     six
     pytest
+    pytest-cov
     html5lib
     mock
     enum34
     typing
 setenv =
     SPHINX_TEST_TEMPDIR = {envdir}/testbuild
-    PYTHONDONTWRITEBYTECODE = true
```
Comment (tk0miya) [inline tox.ini:15]: I noticed this line now. Why did you remove it? I don't opposite for this, but I'd like to know reason.

**Comment by tk0miya** [comment]
@shimizukawa do you have any comment? if none, I will merge this soon.
I think this is good improvement of our tox.ini

File: tox.ini (hunk 1 above)
Comment (stephenfin) [reply tox.ini:15]: From [here](https://github.com/sphinx-doc/sphinx/pull/4117/commits/6086d7b608a365dc3ccfcb37d8bac7f02de338fe):

> As I've previously discovered [1], this doesn't actually do anything in
> modern tox. Simply remove it.
>
> [1] https://review.openstack.org/#/c/369986/

I put all my rationale in my commit messages. Maybe I should start copy-pasting them into the pull request message? :smile:

Comment (tk0miya) [reply tox.ini:15]: I understood. I will read your commit message in next time :-)

**Comment by tk0miya** [comment]
Merged. Thanks!
